Add draft of new browser-based exercise (#11)
trussworks/truss-interview: BROWSER_README.md

_This is one of the steps in the Truss interview process. If you've
stumbled upon this repository and are interested in a career with
Truss, [check out our jobs page](https://truss.works/jobs)._

# Truss Software Engineering Interview

## Introduction and expectations

Hi there! Please complete the problem described below to the best of
your ability, using the tools you're most comfortable with. Assume
you're sending your submission in for code review from peers;
we'll be talking about your submission in your interview in that
context.

We expect this to take less than 4 hours of actual coding time. Please
submit a working but incomplete solution instead of spending more time
on it. We're also aware that getting after-hours coding time can be
challenging; we'd like a submission within a week and if you need more
time please let us know.

Approach this solution the way you would a real world problem. Use
libraries where it makes sense to, and be prepared to explain your
thought process if you do.

If you have any questions, please contact hiring@truss.works; we're
happy to help if you're not sure what we're asking for or if you have
questions.

## How to submit your response

> Please send hiring@truss.works a link to a public git repository

Reviewer: add option to alternatively send a link to your code in an online IDE such as Codepen

(Github is fine) that contains your code and a README.md that tells us
how to build and run it. Your code will be run in Firefox, Chrome, or Safari.

## The problem: Display data from an API for human consumption

Please write a static webpage that loads data from `https://swapi.dev/api/planets/` and displays that data

> in an HTML `table`.

Reviewer: I would not be so explicit about what HTML tags to use (`table, ul, li`). this is another place where I think it will be interesting to see how they interpret it and decide how what markup to use.

> Name the entrypoint for your application `index.html`. To test your solution, we will open `index.html` in a browser.

Reviewer: What if we provide specific info that should encompass all submission-type possibilities? for example:
- If your solution code requires compilation (such as with a tool like webpack or create-react-app), please provide complete instructions, including a package.json file with all dependencies included, for running your code as part of your submission.
- If your solution is in the form of static assets (such as an index.html file) we will run it using a simple no-config HTTP server (https://github.com/vercel/serve)
- If your solution is on a hosted IDE such as Codepen or Glitch, please make sure to provide an evergreen link that does not expire

Author: I like this. It covers all of the bases pretty well.

You may use additional files to organize your code, but please include submit

> a built version of the project that doesn't require the use of any build tools

Reviewer 2: `include submit a built version` ?  Maybe we don't need `submit` ?

submission if you do. Accordingly, you may use any build tool you'd like

> as long as you include files that are ready for viewing directly in a browser.

Reviewer 3: ```suggestion
Name the entrypoint for your application `index.html`. 

You may use additional files to organize your code. You may also use frameworks and build tools as long as clear, simple instructions are provided for how to run. Please ensure your solution can be easily viewed in the browser. Be ready to answer questions about why you chose a given approach.

```

Don't worry about making the table look fancy; you can style it however you
would like to outside the requirements below. We will look at the markup you
generate and any code you write to load or display the data.

## What to display

For each planet in the dataset, please display:

> - The planet's name

Reviewer: Additional requirement suggestion:
> - Link the planet's name to the planet's API URL and opens in a new window or tab

- The planet's climate
- How many residents the planet has
- The terrains found on the planet. Please use an unordered list (`ul`) to structure these and wrap each terrain in a list item (`li`).
- The population
- The surface area covered by water
    - Assume that all planets are perfect spheres.
    - The radius of a sphere is half its diameter.

>     - The `surface_water` value in the data is a percent, so `50` means the planet is 50% covered in water.

Reviewer: minor reword suggestion for clarity:
> The value of `surface_water` from the API data is a percentage, so a value of `50` means the planet is 50% covered in water.

Additionally, please satisfy these requirements:

- Show a loading message while loading the data, and hide this message once the data is displayed.

> - Sort the planets by name, showing earlier letters first.

Reviewer: "alphabetically"?

Author: I knew there was a better way to say that!

- Any column that is unknown should be displayed as `?`.
- Format all numbers larger than 999 with spaces to group digits into groups of three.
  For example, ten thousand should be displayed as `10 000`.

> - Cells in the table should be separated by 1 pixel gray lines.

Reviewer: Do we want them to satisfy an error handling requirement? We could say something like:
> Display an error message if the data load fails for some reason. We may test your code against an invalid URL to demonstrate this.

(alternatively this could be another stretch goal)

Author: I think so. I'll add it as a requirement, and we can make it a stretch goal if we get feedback that it's too much.

## Stretch goals (optional)

If you complete the above and still have additional time, you can choose to complete one or more of the following extra requirements.

> **This is absolutely not required for submitting the work sample.**

Reviewer: add: **and should not be worked on beyond the four hour time limit**

- The API only returns the first page of data by default. Add a "load more" button that
  loads additional pages of data when clicked. Only show this button if there is more data to be loaded.

> - Sort the table by a column's values when that column's header is clicked.

Reviewer: More stretch goal ideas:
- Spruce up your page with some extra styling & design elements
- Make sure your page is usable on mobile devices
- Include some tests

Truss, [check out our jobs page](https://truss.works/jobs)._

# The problem: Display data from an API for human consumption

## Introduction

Approach this solution the way you would a real world problem. Use
libraries where it makes sense to, and be prepared to explain your
thought process if you do. We expect this to take less than 4 hours of actual coding time. Please
submit a working but incomplete solution instead of spending more time
on it.

If you have any questions, please contact [hiring@truss.works](mailto:hiring@truss.works); we're
happy to help.

## How to submit your response

Please submit your solution by emailing [hiring@truss.works](mailto:hiring@truss.works) a link to **one** of the following:

-   A public git repository (Github is fine) that contains your code and a `README.md` that tells us how to build and run it.
-   Your solution running in an online IDE such as [CodePen](https://codepen.io) or [Glitch](https://glitch.com).

Please ensure that your submission includes everything needed to run and view your project:

-   If your solution code requires compilation (such as with a tool like webpack or create-react-app),

>     please provide as a part of your submission complete instructions for running your code, including

Reviewer: ```suggestion
    please provide complete instructions for running your code, including
```

    a `package.json` file with all dependencies included.
-   If your solution is in the form of static assets (such as an index.html file), we will run it
    using a simple no-config HTTP server (<https://github.com/vercel/serve>).
-   If your solution is on a hosted IDE such as Codepen or Glitch, please make sure to
    provide an evergreen link that does not expire.
    
We will review your solution in a recent version of Firefox, Chrome, or Safari.

## Full Description 

> Please write a static webpage that loads data from `https://swapi.dev/api/planets/` and displays that datain an HTML `table`.

Reviewer: ```suggestion
Please write a static webpage that loads data from `https://swapi.dev/api/planets/`, formats and displays that data in a table.
```

Name the entrypoint for your application `index.html`.

You may use additional files to organize your code. You may also use frameworks and build tools as long as clear, simple instructions are provided for how to run them. Please ensure your solution can be easily viewed in the browser. Be ready to answer questions about why you chose a given approach.

Don't worry about making the table look fancy; you can style it however you

> would like to outside the requirements below. We will look at the markup you

Reviewer: ```suggestion
would like to alongside the requirements below. We will look at the markup you
```

Additionally, please satisfy these requirements:

-   Show a loading message while loading the data, and hide this message once the data is displayed.
-   Display an error message if the data load fails for some reason. We may test your code against an invalid URL to demonstrate this.
-   Sort the planets alphabetically
-   Any column that is unknown should be displayed as `?`.
-   Format all numbers larger than 999 with spaces to group digits into groups of three.
    For example, ten thousand should be displayed as `10 000`.
-   Cells in the table should be separated by 1 pixel gray lines.

### Stretch goals (optional)

If you complete the above and still have additional time, you can choose to complete one or more of the following.

**This is absolutely not required for submitting the work sample and should not be worked on beyond the four hour time limit.**

-   The API only returns the first page of data by default. Add a "load more" button that
    loads additional pages of data when clicked. Only show this button if there is more data to be loaded.
-   Sort the table by a column's values when that column's header is clicked.

> -   Table has both an empty state and loading state (when no data exists and when data is still loading accordingly).

Reviewer: ```suggestion
-   Table has both an empty state and loading state (displayed when no data exists, and when data is loading respectively).
```

- Display an error message if the data load fails for some reason. We may test your code against an invalid URL to demonstrate this.
- Sort the planets alphabetically
- Any column that is unknown should be displayed as `?`.
- Format all numbers larger than 999 with spaces to group digits into groups of three.
  For example, ten thousand should be displayed as `10 000`.
- Cells in the table should be separated by 1 pixel gray lines.

## Stretch goals (optional)

If you complete the above and still have additional time, you can choose to complete one or more of the following extra requirements.

**This is absolutely not required for submitting the work sample and should not be worked on beyond the four hour time limit.**

- The API only returns the first page of data by default. Add a "load more" button that
  loads additional pages of data when clicked. Only show this button if there is more data to be loaded.
- Sort the table by a column's values when that column's header is clicked.
- Spruce up your page with some extra styling & design elements!
- Make sure your page is usable on mobile devices.

> - Include some tests.

Reviewer 3: - Table has both an empty state and loading state (when no data exists and when data is still loading accordingly).

time please let us know.

Approach this solution the way you would a real world problem. Use
libraries where it makes sense to, and be prepared to explain your
thought process if you do.

We will review your solution in a recent version of Firefox, Chrome, or Safari.

If you have any questions, please contact [hiring@truss.works](mailto:hiring@truss.works); we're
happy to help if you're not sure what we're asking for or if you have
questions.

## How to submit your response

Please submit your solution by emailing [hiring@truss.works](mailto:hiring@truss.works) a link to **one** of the following:

-   A public git repository (Github is fine) that contains your code and a `README.md` that tells us how to build and run it.
-   Your solution running in an online IDE such as [CodePen](https://codepen.io) or [Glitch](https://glitch.com).

Please ensure that your submission includes everything needed to run and view your project:

-   If your solution code requires compilation (such as with a tool like webpack or create-react-app),
    please provide as a part of your submission complete instructions for running your code, including

>     a `package.json` file with all dependencies included.

Reviewer 4: Will these instructions will vary by platform? For example if a candidate is developing on linux but we are running their code on macOS, will the supplied instructions for installing dependencies work?

Reviewer: it is probably worth saying we'll be running their solutions on MacOS but they can specify a node version for example

-   If your solution is in the form of static assets (such as an index.html file), we will run it
    using a simple no-config HTTP server (<https://github.com/vercel/serve>).
-   If your solution is on a hosted IDE such as Codepen or Glitch, please make sure to

>     provide an evergreen link that does not expire.

Reviewer 4: What is an evergreen link?

Reviewer: basically one that will represent the correct state of their code at the time of submission, won't expire or change to something else, ie if they continue to make edits. there's probably a better way to phrase that and I'm also not sure what these platforms support

## The problem: Display data from an API for human consumption

Please write a static webpage that loads data from `https://swapi.dev/api/planets/` and displays that data
in an HTML `table`.

Name the entrypoint for your application `index.html`.

> You may use additional files to organize your code. You may also use frameworks and build tools as long as clear, simple instructions are provided for how to run. Please ensure your solution can be easily viewed in the browser. Be ready to answer questions about why you chose a given approach.

Reviewer 4: ```suggestion
You may use additional files to organize your code. You may also use frameworks and build tools as long as clear, simple instructions are provided for how to run them. Please ensure your solution can be easily viewed in the browser. Be ready to answer questions about why you chose a given approach.
```

-   The planet's name
    -   The name should be a link that, when clicked, opens the planet's API URL in a new window
-   The planet's climate
-   How many residents the planet has
-   The terrains found on the planet
-   The population
-   The surface area covered by water
    -   Assume that all planets are perfect spheres.
    -   The radius of a sphere is half its diameter.
    -   The value of `surface_water` from the API is a percentage, so a value of `50` means the planet is 50% covered in water.
    -   Round these values to the nearest integer value.

Additionally, please satisfy these requirements:

-   Show a loading message while loading the data, and hide this message once the data is displayed.
-   Display an error message if the data load fails for some reason. We may test your code against an invalid URL to demonstrate this.
-   Sort the planets alphabetically
-   Any column that is unknown should be displayed as `?`.
-   Format all numbers larger than 999 with spaces to group digits into groups of three.
    For example, ten thousand should be displayed as `10 000`.
-   Cells in the table should be separated by 1 pixel gray lines.

> ### Stretch goals (optional)

Reviewer 5: Do we want to list stretch goals, or just mention they can keep going and give some examples and see what folks come up with on their own? I'd be curious what they think to do if they have more time vs checking things off a list. 
That said, I'd also be curious what stretch goals they do implement from the list since that could also be insightful.

Reviewer: Hm, I do like the idea of having them come up with their own ideas. Maybe if we specify that the stretch goals are optional, and these are just some ideas but they can take it where they want?

If you complete the above and still have additional time, you can choose to complete one or more of the following extra requirements.

**This is absolutely not required for submitting the work sample and should not be worked on beyond the four hour time limit.**

-   The API only returns the first page of data by default. Add a "load more" button that
    loads additional pages of data when clicked. Only show this button if there is more data to be loaded.
-   Sort the table by a column's values when that column's header is clicked.
-   Table has both an empty state and loading state (when no data exists and when data is still loading accordingly).
-   Spruce up your page with some extra styling & design elements!
-   Make sure your page is usable on mobile devices.

> -   Include some tests.

Reviewer 4: Making this a stretch goal might actually discourage someone from writing tests although they would have written tests had they never been asked. It is useful to see if an engineer writes tests even when it's not explicitly required (or explicitly optional, for that matter).

Reviewer: the reason for making this a stretch goal (in my opinion) is that there is no native test runner for browser code, so unless you have already decided to use a JS framework, there will be a not insignificant amount of setup to get tests running.
